File: hellogo/pkg/chess/common/perft_result_test.go
package chesscommon

import (
	"testing"
)

func TestNewPerftResult(t *testing.T) {
	result := NewPerftResult()
	if result.LeafNodesCount() != 0 {
		t.Errorf("Expected default leafNodesCount to be 0, got %d", result.LeafNodesCount())
	}
	if result.SearchedNodesCount() != 0 {
		t.Errorf("Expected default searchedNodesCount to be 0, got %d", result.SearchedNodesCount())
	}
}

func TestSettersAndGetters(t *testing.T) {
	result := NewPerftResult()

	// Test SetLeafNodesCount and LeafNodesCount
	result.SetLeafNodesCount(42)
	if got := result.LeafNodesCount(); got != 42 {
		t.Errorf("Expected leafNodesCount to be 42, got %d", got)
	}

	// Test IncrementSearchedNodesCount and SearchedNodesCount
	result.IncrementSearchedNodesCount()
	if got := result.SearchedNodesCount(); got != 1 {
		t.Errorf("Expected searchedNodesCount to be 1 after increment, got %d", got)
	}

	result.IncrementSearchedNodesCount()
	if got := result.SearchedNodesCount(); got != 2 {
		t.Errorf("Expected searchedNodesCount to be 2 after second increment, got %d", got)
	}
}

func TestEquals(t *testing.T) {
	tests := []struct {
		name     string
		a        *PerftResult
		b        *PerftResult
		expected bool
	}{
		{"both nil", nil, nil, true},
		{"a nil", nil, NewPerftResult(), false},
		{"b nil", NewPerftResult(), nil, false},
		{"both default", NewPerftResult(), NewPerftResult(), true},
		{"same values", createPerftResult(10, 20), createPerftResult(10, 20), true},
		{"different leaf nodes", createPerftResult(10, 20), createPerftResult(11, 20), false},
		{"different searched nodes", createPerftResult(10, 20), createPerftResult(10, 21), false},
		{"different both", createPerftResult(10, 20), createPerftResult(11, 21), false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			var result bool
			if tt.a == nil {
				result = (tt.b == nil) == tt.expected
			} else {
				result = tt.a.Equals(tt.b) == tt.expected
			}

			if !result {
				t.Errorf("Equals() failed for %s, expected %v", tt.name, tt.expected)
			}

			// Test symmetry
			if tt.a != nil && tt.b != nil {
				if tt.a.Equals(tt.b) != tt.b.Equals(tt.a) {
					t.Errorf("Equals() is not symmetric for %s", tt.name)
				}
			}
		})
	}
}

func TestHashCode(t *testing.T) {
	tests := []struct {
		a        *PerftResult
		b        *PerftResult
		sameHash bool
	}{
		{NewPerftResult(), NewPerftResult(), true},
		{createPerftResult(10, 20), createPerftResult(10, 20), true},
		{createPerftResult(10, 20), createPerftResult(11, 20), false},
		{createPerftResult(10, 20), createPerftResult(10, 21), false},
	}

	for _, tt := range tests {
		t.Run("", func(t *testing.T) {
			hashA := tt.a.HashCode()
			hashB := tt.b.HashCode()

			if tt.sameHash && hashA != hashB {
				t.Errorf("Expected equal hash codes, got %d and %d", hashA, hashB)
			} else if !tt.sameHash && hashA == hashB {
				// This is not strictly required (hash collisions can happen),
				// but it's good to test for common cases
				t.Logf("Different objects have same hash code: %d and %d", hashA, hashB)
			}
		})
	}
}

func TestEqualsAndHashCodeConsistency(t *testing.T) {
	r1 := createPerftResult(10, 20)
	r2 := createPerftResult(10, 20)

	if r1.Equals(r2) && r1.HashCode() != r2.HashCode() {
		t.Error("Equal objects must have equal hash codes")
	}
}

// Helper function to create a PerftResult with specific values
func createPerftResult(leafNodes, searchedNodes int64) *PerftResult {
	result := NewPerftResult()
	result.SetLeafNodesCount(leafNodes)
	for i := int64(0); i < searchedNodes; i++ {
		result.IncrementSearchedNodesCount()
	}
	return result
}

File: hellojava/src/test/java/com/fathzer/chess/common/PerftResultTest.java
package com.fathzer.chess.common;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class PerftResultTest {

    @Test
    void testDefaultConstructor() {
        PerftResult result = new PerftResult();
        assertEquals(0, result.leafNodesCount(), "Default leafNodesCount should be 0");
        assertEquals(0, result.searchedNodesCount(), "Default searchedNodesCount should be 0");
    }

    @Test
    void testSettersAndGetters() {
        PerftResult result = new PerftResult();
        
        // Test setLeafNodesCount and leafNodesCount
        result.setLeafNodesCount(42);
        assertEquals(42, result.leafNodesCount(), "leafNodesCount should be 42");
        
        // Test incrementSearchedNodesCount and searchedNodesCount
        result.incrementSearchedNodesCount();
        assertEquals(1, result.searchedNodesCount(), "searchedNodesCount should be 1 after increment");
        
        result.incrementSearchedNodesCount();
        assertEquals(2, result.searchedNodesCount(), "searchedNodesCount should be 2 after second increment");
    }

    @ParameterizedTest
    @CsvSource({
        "0, 0, 0, 0, true",
        "10, 20, 10, 20, true",
        "10, 20, 10, 10, false",
        "10, 20, 20, 20, false",
        "10, 20, 20, 10, false"
    })
    void testEquals(long leaf1, long searched1, long leaf2, long searched2, boolean expected) {
        // Create first PerftResult
        PerftResult result1 = createPerftResult(leaf1, searched1);
        
        // Create second PerftResult
        PerftResult result2 = createPerftResult(leaf2, searched2);
        
        // Test equals
        assertEquals(expected, result1.equals(result2), "equals() returned unexpected result");
        
        // Test equals with null
        assertFalse(result1.equals(null), "equals(null) should return false");
        
        // Test equals with different class
        assertFalse(result1.equals("not a PerftResult"), "equals() with different class should return false");
        
        // Test reflexive
        assertEquals(result1, result1, "equals() should be reflexive");
        
        // Test symmetric
        if (expected) {
            assertEquals(result2, result1, "equals() should be symmetric");
        } else {
            assertNotEquals(result2, result1, "equals() should be symmetric");
        }
    }

    @ParameterizedTest
    @CsvSource({
        "0, 0, 0, 0, true",
        "10, 20, 10, 20, true",
        "10, 20, 10, 10, false",
        "10, 20, 20, 20, false"
    })
    void testHashCode(long leaf1, long searched1, long leaf2, long searched2, boolean expectedSameHash) {
        // Create first PerftResult
        PerftResult result1 = createPerftResult(leaf1, searched1);
        
        // Create second PerftResult
        PerftResult result2 = createPerftResult(leaf2, searched2);
        
        if (expectedSameHash) {
            assertEquals(result1.hashCode(), result2.hashCode(), 
                "hashCode() should be equal for equal objects");
        } else {
            // Note: Not strictly required for hash codes to be different for unequal objects,
            // but it's good practice to test for common cases
            assertNotEquals(result1.hashCode(), result2.hashCode(),
                "hashCode() should be different for different objects");
        }
    }

    @Test
    void testEqualsAndHashCodeConsistency() {
        PerftResult result1 = createPerftResult(10, 20);
        PerftResult result2 = createPerftResult(10, 20);
        
        // If two objects are equal, their hash codes must be equal
        if (result1.equals(result2)) {
            assertEquals(result1.hashCode(), result2.hashCode(),
                "Equal objects must have equal hash codes");
        }
    }

    // Helper method to create a PerftResult with specific values
    private PerftResult createPerftResult(long leafNodes, long searchedNodes) {
        PerftResult result = new PerftResult();
        result.setLeafNodesCount(leafNodes);
        for (long i = 0; i < searchedNodes; i++) {
            result.incrementSearchedNodesCount();
        }
        return result;
    }
}
